### PR TITLE
Issue 1936: Ensure system test framework is compatible with Marathon version 1.5.0 DC/OS version: 1.10.0

### LINKED ISSUE #1936: System test framework should work with the latest version of Marathon.
**Problem description**
Marathon apis used by System test framework have been deprecated with the Marathon version 1.5.x.
This causes a failure to deploy services on the latest mesos cluster.

**Problem location**
System test framework.

**Suggestions for an improvement**
Ensure the latest apis are used.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**
* Ensure marathon-client compatible with Marathon v1.5.0 is used.
* The latest version of marathon does not accept port as array of integers. It returns http code 422 (Unprocessable entity) for ports are passed as an array of integers. This causes failure to deploy services on the mesos/dcos cluster.

**Purpose of the change**
Fixes #1936

**What the code does**
- Upgrades the client library used.
- Ensure PortDefnitions objects are used instead of array of ports.

**How to verify it**
Verified that services are getting deployed and tests are getting executed.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -485,7 +486,7 @@ project('test:system') {
     dependencies {
         // https://mvnrepository.com/artifact/com.mesosphere/marathon-client
         runtime group: 'ch.qos.logback', name: 'logback-classic', version: qosLogbackVersion
-        compile group: 'com.mesosphere', name: 'marathon-client', version: marathonClientVersion
+        compile group: 'com.mesosphere', name: 'marathon-client', version: marathonClientVersion, withoutSlf4jSimple
```
Comment (fpj) [inline build.gradle]: Could you explain this change, please?

Comment (shrids) [reply build.gradle]: The latest marathon-client v0.6.0 has a transitive dependency on group: 'org.slf4j', module: 'slf4j-simple'. We use logback as our slf4j implementation. Due to the transitive dependency we end up have both the implementations in the class path causing logging to not work.

**Comment by shrids** [comment]
Apart from upgrading the marathonclient library from 0.3.0 to 0.6.0 we also need to update the various service implementations to ensure service deployment works, specifically the portdefnition change.
 
Only if we are using marathon 1.x should we use 0.6.0 version of the marathon client.

File: build.gradle [hunk 2]
```diff
@@ -72,6 +72,7 @@ project('common') {
 }
 
 def withoutLog4j = { exclude group: 'org.slf4j', module: 'slf4j-log4j12' }
+def withoutSlf4jSimple = { exclude group: 'org.slf4j', module: 'slf4j-simple' }
```
Comment (tkaitchuck) [inline build.gradle]: I think this can probably be combined with the above, and can be called withoutLogger. As that's essentially what we are overriding.

Comment (shrids) [reply build.gradle]: done.

File: test/system/src/main/java/io/pravega/test/system/framework/marathon/AuthEnabledMarathonClient.java [hunk 3]
```diff
@@ -21,10 +21,11 @@
 import feign.gson.GsonEncoder;
 import io.pravega.test.system.framework.LoginClient;
 import java.util.Calendar;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/main/java/io/pravega/test/system/framework/marathon/AuthEnabledMarathonClient.java]: unnecessary new line.

File: test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java [hunk 4]
```diff
@@ -128,10 +128,10 @@ private App createPravegaControllerApp() {
         parameterList.add(element1);
         app.getContainer().getDocker().setParameters(parameterList);
         //set port
-        app.setPorts(Arrays.asList(CONTROLLER_PORT, REST_PORT));
+        app.setPortDefinitions(Arrays.asList(createPortDefinition(CONTROLLER_PORT), createPortDefinition(REST_PORT)));
         app.setRequirePorts(true);
         List<HealthCheck> healthCheckList = new ArrayList<HealthCheck>();
-        healthCheckList.add(setHealthCheck(900, "TCP", false, 60, 20, 0));
+        healthCheckList.add(setHealthCheck(900, "TCP", false, 60, 20, 0, CONTROLLER_PORT));
```
Comment (Prabha-Veerubhotla) [inline test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java]: don't  we need `REST-PORT` also here?

File: test/system/src/main/java/io/pravega/test/system/framework/services/MarathonBasedService.java [hunk 5]
```diff
@@ -71,7 +73,9 @@ public boolean isRunning() {
             GetAppResponse app = marathonClient.getApp(this.id);
             log.debug("App Details: {}", app);
             //app is not running until the desired instance count is equal to the number of task/docker containers
-            if (app.getApp().getTasksRunning().intValue() == app.getApp().getInstances().intValue()) {
+            // and the state of application is healthy.
+            if ((app.getApp().getTasksRunning().intValue() == app.getApp().getInstances().intValue())
+                    && app.getApp().getTasksRunning().intValue() == app.getApp().getTasksHealthy().intValue()) {
```
Comment (Prabha-Veerubhotla) [inline test/system/src/main/java/io/pravega/test/system/framework/services/MarathonBasedService.java]: Will getTasksHealthy() alone be sufficient ?

File: test/system/src/test/java/io/pravega/test/system/ControllerFailoverTest.java [hunk 6]
```diff
@@ -34,7 +34,8 @@
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/ControllerFailoverTest.java]: unnecessary new line

File: test/system/src/test/java/io/pravega/test/system/BookkeeperTest.java [hunk 7]
```diff
@@ -17,7 +17,8 @@
 import java.net.URI;
 import java.util.List;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/BookkeeperTest.java]: unnecessary new line

File: test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java [hunk 8]
```diff
@@ -28,7 +28,8 @@
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java]: unnecessary new line

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java [hunk 9]
```diff
@@ -31,7 +31,8 @@
 import java.util.Random;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: new line

File: test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java [hunk 10]
```diff
@@ -31,7 +31,8 @@
 import java.util.Random;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java]: unnecessary new line

File: test/system/src/test/java/io/pravega/test/system/MultiSegmentStoreTest.java [hunk 11]
```diff
@@ -36,7 +36,8 @@
 import java.util.UUID;
 import lombok.Cleanup;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/MultiSegmentStoreTest.java]: check the  above comment.

File: test/system/src/test/java/io/pravega/test/system/PravegaControllerTest.java [hunk 12]
```diff
@@ -17,7 +17,8 @@
 import java.net.URI;
 import java.util.List;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/PravegaControllerTest.java]: check the above comment.

File: test/system/src/test/java/io/pravega/test/system/PravegaSegmentStoreTest.java [hunk 13]
```diff
@@ -19,7 +19,8 @@
 import java.net.URI;
 import java.util.List;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/PravegaSegmentStoreTest.java]: new line.

File: test/system/src/test/java/io/pravega/test/system/PravegaTest.java [hunk 14]
```diff
@@ -40,7 +40,8 @@
 import java.util.concurrent.ExecutionException;
 import lombok.Cleanup;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/PravegaTest.java]: new line.

File: test/system/src/test/java/io/pravega/test/system/ReadTxnWriteAutoScaleWithFailoverTest.java [hunk 15]
```diff
@@ -33,7 +33,8 @@
 import java.util.Random;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/ReadTxnWriteAutoScaleWithFailoverTest.java]: new line.

File: test/system/src/test/java/io/pravega/test/system/ReadWriteAndAutoScaleWithFailoverTest.java [hunk 16]
```diff
@@ -32,7 +32,8 @@
 import java.util.Random;
 import java.util.stream.Collectors;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/ReadWriteAndAutoScaleWithFailoverTest.java]: check the above comment.

File: test/system/src/test/java/io/pravega/test/system/ZookeeperTest.java [hunk 17]
```diff
@@ -15,7 +15,8 @@
 import io.pravega.test.system.framework.services.ZookeeperService;
 import java.net.URI;
 import lombok.extern.slf4j.Slf4j;
-import mesosphere.marathon.client.utils.MarathonException;
+import mesosphere.marathon.client.MarathonException;
+
```
Comment (Prabha-Veerubhotla) [inline test/system/src/test/java/io/pravega/test/system/ZookeeperTest.java]: new line.

File: test/system/src/main/java/io/pravega/test/system/framework/marathon/AuthEnabledMarathonClient.java (hunk 3 above)
Comment (shrids) [reply test/system/src/main/java/io/pravega/test/system/framework/marathon/AuthEnabledMarathonClient.java]: removed it.

File: test/system/src/main/java/io/pravega/test/system/framework/services/MarathonBasedService.java (hunk 5 above)
Comment (shrids) [reply test/system/src/main/java/io/pravega/test/system/framework/services/MarathonBasedService.java]: yes, we are ensuring that the taskrunning.count and task.healthy.count matches the desired instance.count.

File: test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java (hunk 4 above)
Comment (shrids) [reply test/system/src/main/java/io/pravega/test/system/framework/services/PravegaControllerService.java]: By default marathon uses `portIndex ` of value `0` . The older code used to use portIndex 0, the current code explicitly ensures this grpc port which is used for all failover tests as part of the health check.

File: test/system/src/test/java/io/pravega/test/system/BookkeeperTest.java (hunk 7 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/BookkeeperTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/ControllerFailoverTest.java (hunk 6 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/ControllerFailoverTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java (hunk 8 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/MultiControllerTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java (hunk 9 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/MultiReaderTxnWriterWithFailoverTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java (hunk 10 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/MultiSegmentStoreTest.java (hunk 11 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/MultiSegmentStoreTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/PravegaControllerTest.java (hunk 12 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/PravegaControllerTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/PravegaSegmentStoreTest.java (hunk 13 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/PravegaSegmentStoreTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/PravegaTest.java (hunk 14 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/PravegaTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/ReadTxnWriteAutoScaleWithFailoverTest.java (hunk 15 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/ReadTxnWriteAutoScaleWithFailoverTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/ReadWriteAndAutoScaleWithFailoverTest.java (hunk 16 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/ReadWriteAndAutoScaleWithFailoverTest.java]: removed it.

File: test/system/src/test/java/io/pravega/test/system/ZookeeperTest.java (hunk 17 above)
Comment (shrids) [reply test/system/src/test/java/io/pravega/test/system/ZookeeperTest.java]: removed it.

**Pull request merged**